crates/sui-indexer/src/store/pg_partition_manager.rs

```diff
@@ -0,0 +1,139 @@
+    fn from(epoch: EpochToCommit) -> Self {
+        let last_epoch = epoch.last_epoch.as_ref().map(|e| e.epoch).unwrap_or(0);
+        let next_epoch = epoch.new_epoch.epoch;
+        let last_epoch_start_cp = epoch.last_epoch.map(|e| e.first_checkpoint_id).unwrap_or(0);
+        let next_epoch_start_cp = epoch.new_epoch.first_checkpoint_id;
+        Self {
+            last_epoch,
+            next_epoch,
+            last_epoch_start_cp,
+            next_epoch_start_cp,
+        }
+    }
+}
+
+impl PgPartitionManager {
+    pub fn new(cp: PgConnectionPool) -> Result<Self, IndexerError> {
+        let manager = Self { cp };
+        let tables = manager.get_table_partitions()?;
+        info!(
+            "Found {} tables with partitions : [{:?}]",
+            tables.len(),
+            tables
+        );
+        Ok(manager)
+    }
+
+    pub fn get_table_partitions(&self) -> Result<BTreeMap<String, u64>, IndexerError> {
+        #[derive(QueryableByName, Debug, Clone)]
+        struct PartitionedTable {
+            #[diesel(sql_type = VarChar)]
+            table_name: String,
+            #[diesel(sql_type = VarChar)]
+            last_partition: String,
+        }
+
+        Ok(
+            read_only_blocking!(&self.cp, |conn| diesel::RunQueryDsl::load(
+                diesel::sql_query(GET_PARTITION_SQL),
+                conn
+            ))?
+            .into_iter()
+            .map(|table: PartitionedTable| {
+                u64::from_str(&table.last_partition)
+                    .map(|last_partition| (table.table_name, last_partition))
+                    .map_err(|e| anyhow!(e))
+            })
+            .collect::<Result<_, _>>()?,
+        )
+    }
+
+    pub fn advance_table_epoch_partition(
+        &self,
+        table: &str,
+        data: &EpochPartitionData,
+    ) -> Result<(), IndexerError> {
+        let (last_epoch, next_epoch, last_epoch_start_cp, next_epoch_start_cp) = (
+            data.last_epoch,
+            data.next_epoch,
+            data.last_epoch_start_cp,
+            data.next_epoch_start_cp,
+        );
+        // first partition has been created in /migrations
+        if next_epoch == 0 {
+            tracing::warn!("skipping epoch 0");
+            return Ok(());
+        }
+        let detach_last_partition =
+            format!("ALTER TABLE {table} DETACH PARTITION {table}_partition_{last_epoch};");
+        transactional_blocking!(&self.cp, |conn| {
+            RunQueryDsl::execute(diesel::sql_query(detach_last_partition), conn)
+        })?;
+
+        let create_new_partition = format!(
+            "CREATE TABLE {table}_partition_{next_epoch} PARTITION OF {table}
+            FOR VALUES FROM ({next_epoch_start_cp}) TO (MAXVALUE);"
+        );
+        let move_rows_to_new_partition = format!(
+            "INSERT INTO {table}_partition_{next_epoch} SELECT * FROM {table}_partition_{last_epoch} WHERE checkpoint_sequence_number >= {next_epoch_start_cp};"
+        );
```

Comment:
can we avoid copying data during new partition creation, and make sure the data always goes to the right partition?

Reply:
sry I am not following, there is always constant write traffic, meaning that we will expect newest rows to go to a "wrong" partition before the corresponding partition is created.